File: leetcode/trie.py
#!/usr/bin/env python3

import sys
import unittest

# number: 208
# section: trie (prefix tree)
# difficulty: medium
# tags: hash table, string, design, trie, top 150, citadel

# constraints
# 1 <= word.length, prefix.length <= 2000
# word and prefix consist only of lowercase English letters.
# At most 3 * 10^4 calls in total will be made to insert, search, and startsWith.

# complexity
# run-time: O(n)
# space: O(n*m)
class Trie:

    def __init__(self):
        self.children = {}
        self.isendword = False

    def insert(self, word: str) -> None:
        node = self

        for c in word:
            if c not in node.children:
                node.children[c] = Trie()
            node = node.children[c]

        node.isendword = True

    def search(self, word: str) -> bool:
        return self.__search_helper(word, True)

    def starts_with(self, prefix: str) -> bool:
        return self.__search_helper(prefix)

    def __search_helper(self, word: str, check_end_word=False) -> bool:
        node = self

        for c in word:
            if c not in node.children:
                return False
            node = node.children[c]

        return True if not check_end_word or node.isendword else False

class TestTrie(unittest.TestCase):
    def test(self):
        trie = Trie()
        trie.insert("apple")
        self.assertTrue(trie.search("apple"))
        self.assertFalse(trie.search("app"))
        self.assertTrue(trie.starts_with("app"))
        trie.insert("app")
        self.assertTrue(trie.search("app"))

if __name__ == '__main__':
    sys.exit(unittest.main())

File: leetcode/trie.cpp
#include <map>
#include <string>
#include <utility>

// number: 208
// title: Implement Trie (Prefix Tree)
// url: https://leetcode.com/problems/implement-trie-prefix-tree/
// section: trie (prefix tree)
// difficulty: medium
// tags: hash table, string, design, trie, top 150, citadel

// constraints
// 1 <= word.length, prefix.length <= 2000
// word and prefix consist only of lowercase English letters.
// At most 3 * 10^4 calls in total will be made to insert, search, and
// startsWith.

// complexity
// run-time: O(n)
// space: O(n*m)
class Trie {
public:
    using ChildrenMap = std::map<char, Trie*>;

    Trie() = default;

    void insert(const std::string& word) {
        Trie* curr = this;

        for (const auto& c: word) {
            ChildrenMap& children = curr->d_children;
            auto itr = children.find(c);
            if (itr == children.end()) {
                children[c] = new Trie();
                // if we had to move it
                //Trie newTrie = Trie();
                //children.insert(std::make_pair(c, std::move(newTrie)));
            }

            curr = curr->d_children[c];
        }

        curr->isendword(true);
    }

    bool search(const std::string& word) {
        return searchHelper(word, true);
    }

    bool startsWith(const std::string& prefix) {
        return searchHelper(prefix);
    }

    void isendword(bool val) { d_isendword = val; }

    bool isendword() const { return d_isendword; }

private:
    bool searchHelper(const std::string& word, bool checkEndWord=false) {
        Trie* curr = this;

        for (const auto& c: word) {
            // can access private members of other instances of same class!
            ChildrenMap& children = curr->d_children;
            const auto itr = children.find(c);
            if (itr == children.end()) {
                return false;
            }

            curr = curr->d_children[c];
        }

        if (!checkEndWord || curr->isendword()) {
            return true;
        } else {
            return false;
        }
    }

    ChildrenMap d_children;
    bool        d_isendword = false;
};

// Your Trie object will be instantiated and called as such:
Trie* obj = new Trie();
obj->insert(word);
bool param_2 = obj->search(word);
bool param_3 = obj->startsWith(prefix);
